akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala

```diff
@@ -778,19 +778,21 @@ private[akka] class ShardRegion(
   }
 
   def receiveCommand(cmd: ShardRegionCommand): Unit = cmd match {
-    case Retry =>
+
+    case Retry if gracefulShutdownInProgress =>
       sendGracefulShutdownToCoordinator()
+      tryCompleteGracefulShutdown()
 
+    case Retry =>
       if (shardBuffers.nonEmpty)
         retryCount += 1
+
```

Comment:
From what I see here, the keep sending a `Retry` tick to the region, but it has two purposes. 

If it's shutting down, we want to send another `GracefulShutdownReq` to the coordinator and then try to complete the shutdown. 

If not shutting down, the are two possible outcomes. 

Either we still need to register to coordinator, or we need to get the shard homes for the messages we have buffered. 

I think it make sense to split it into two distinctive case (could be an `if/else` under `case Retry`). 

Please, let me know if that makes sense and if not, clarify why. ;-)

Reply:
I think it's more of a periodic tick that was perhaps somewhat misleadingly named retry, because it triggers retries of different things (retry register to coordinator, retry request sharded buffer homes, retry graceful shutdown).

A bit surprising to me that we retry registration both on Retry and on RegisterRetry which is scheduled with does some clever exponential backoff.

If you change this you should probably also clean up the ifs in `sendGracefulShutdownToCoordinator` and `tryCompleteGracefulShutdown` if placing a if-else higher up or it becomes a bit confusing (and add one in `receiveTerminated` as well then).

Reply:
I don't know if I'm following you, but I forgot to do some more clean-up (juggling around with too many things in my head). It turns out that `sendGracefulShutdownToCoordinator` and `tryCompleteGracefulShutdown` have themselves a `if` checking for `gracefulShutdownInProgress`.  

We should clean-up it, IMO. Too many boolean logic spread. 

Anyway, I will do another round and see what can be improved further.

Reply:
And indeed, the `Retry` is misleading. I got confused in the beginning. Then I understood that is like a heartbeat.

Reply:
@johanandren, I did some more changes. 

I'm inclined to add to distinct ticks: one for the GracefulShutdown that is added on the first shutdown message and keep going until it succeed and one for the buffered messages. 

I didn't want to change too much for now. I need to analyse it better, but here is what I can tell for now: 

Instead of check if `coordinator.isEmpty`, we should only check if `shardBuffers.nonEmpty`. 
So, 

```scala
if (shardBuffers.nonEmpty) {
  requestShardBufferHomes()
} else {
	retryCount = 0 
}
```

The `retryCount += 1` should be inside `requestShardBufferHomes`. 

That's the only place we use it. Each time we request the homes, we increase it. And each time we get the heartbeat, we set it to zero if there are not buffered messages. As it is now it increases forever.

Reply:
And I have also found this:

https://github.com/akka/akka/blob/cd202e40a9d76c2e7885038666ac087adb3582b5/akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala#L927-L933

We stop to schedule the `RegisterRetry` if we reach `retryInterval`. 

I think it should be:

```scala
if (nextRegistrationDelay < retryInterval) {
  // exponentially increasing retry interval until reaching the normal retryInterval
  nextRegistrationDelay *= 2
}

val minDelay = math.min(nextRegistrationDelay, retryInterval)
timers.startSingleTimer(RegisterRetry, RegisterRetry, minDelay)
```

Reply:
Isn't the shard coordinator retry on `Retry` there in case the shard coordinator moved to a new node?

I think I'd be more careful with changing many things here at the same time as it is easy to miss some aspect.

Reply:
Yes, it is. Also, I digged in some old PRs. The reason why we have `Retry` and `RegisterRetry` is that the `Retry` was too slow https://github.com/akka/akka/issues/25191. 

The `RegisterRetry` was introduced as a speed-up workaround with an exponential increase. After reach `retryInterval` we let it go and rely on the original `Retry`.